File: fp-scala/src/main/scala/com/cscie88a/week2/Dog.scala
package com.cscie88a.week2

final case class Dog(
  override val name: String,
  override val eyeColor: String,
  override val bodyColor: String
  ) extends Animal(name, eyeColor, bodyColor)
       with Trainable with Guard {

  override def doTrick(trickName: String): Boolean = {
    println(
      s"${name} says - dogs love doing tricks. I'm doing a trick ${trickName} now"
    )
    true
  }

  override def playWithToy(toy: Toy): Boolean = {
    println("I love the toy! Yeah!!!")
    toy.doFunStuff()
    true
  }

  override def singWithFriend(aFriend: Animal): Boolean = {
    println(s"I am singing with my friend ${aFriend.name}")
    true
  }

  override def guardHome(): Boolean = {
    true
  }
}

File: fp-java/src/main/java/cscie88a/week2/Dog.java
package cscie88a.week2;

public class Dog extends Animal implements ITrainable, Guard {

	public Dog() {
	}

	public Dog(String name, String eyeColor, String bodyColor) {
		super(name, eyeColor, bodyColor);
	}

	@Override
	public boolean doTrick(String trickName) {
		System.out.println(name + " says: I LOVE doing tricks! I'm doing "
				+ trickName + " now!");
		return true;
	}

	@Override
	public boolean singWithFriend(Animal aFriend) {
		System.out.println("I am singing with my friend " + aFriend.getName());
		return true;
	}

	@Override
	public boolean playWithToy(Toy toy) {
		System.out.println("I love the toy! Yeah!!!");
		toy.doFunStuff();
		return true;
	}

	@Override
	public boolean guardHome() {
		return true;
	}
}
